File: 01-arrays-and-strings/01_two_sum/TwoSum.java
/**
 * Problem: Two Sum
 * LeetCode: #1 (Easy)
 * 
 * Approach: Hash Map lookup (Single Pass)
 * Time Complexity: O(N)
 * Space Complexity: O(N)
 */

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class TwoSum {
    public static int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (map.containsKey(complement)) {
                return new int[] { map.get(complement), i };
            }
            map.put(nums[i], i);
        }
        return new int[] {};
    }

    public static void main(String[] args) {
        System.out.println("[Java] Two Sum Test");
        int[] nums = { 2, 7, 11, 15 };
        int target = 9;

        int[] res = twoSum(nums, target);
        System.out.println("Array: " + Arrays.toString(nums) + ", Target: " + target);
        System.out.println("Result Indices: " + Arrays.toString(res));
    }
}

File: 01-arrays-and-strings/01_two_sum/two_sum.py
"""
Problem: Two Sum
LeetCode: #1 (Easy)

Approach: Hash Map lookup (Single Pass)
Time Complexity: O(N)
Space Complexity: O(N)
"""

from typing import List

def two_sum(nums: List[int], target: int) -> List[int]:
    seen = {}
    for i, num in enumerate(nums):
        complement = target - num
        if complement in seen:
            return [seen[complement], i]
        seen[num] = i
    return []

if __name__ == "__main__":
    print("[Python] Two Sum Test")
    tests = [
        ([2, 7, 11, 15], 9, [0, 1]),
        ([3, 2, 4], 6, [1, 2]),
        ([3, 3], 6, [0, 1]),
    ]
    for nums, target, expected in tests:
        res = two_sum(nums, target)
        print(f"nums: {nums}, target: {target} -> Result: {res}")
        assert res == expected, f"Failed for {nums}"
    print("All Two Sum tests passed!")
